presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayFilterFunction.java

```diff
@@ -0,0 +1,106 @@
+import com.facebook.presto.metadata.Signature;
+import com.facebook.presto.metadata.SqlScalarFunction;
+import com.facebook.presto.spi.block.Block;
+import com.facebook.presto.spi.block.BlockBuilder;
+import com.facebook.presto.spi.block.BlockBuilderStatus;
+import com.facebook.presto.spi.type.Type;
+import com.facebook.presto.spi.type.TypeManager;
+import com.google.common.base.Throwables;
+import com.google.common.collect.ImmutableList;
+
+import java.lang.invoke.MethodHandle;
+
+import static com.facebook.presto.metadata.Signature.typeVariable;
+import static com.facebook.presto.spi.type.TypeSignature.parseTypeSignature;
+import static com.facebook.presto.spi.type.TypeUtils.readNativeValue;
+import static com.facebook.presto.util.Reflection.methodHandle;
+import static java.lang.Boolean.TRUE;
+
+public final class ArrayFilterFunction
+        extends SqlScalarFunction
+{
+    public static final ArrayFilterFunction ARRAY_FILTER_FUNCTION = new ArrayFilterFunction();
+
+    private static final MethodHandle METHOD_HANDLE = methodHandle(ArrayFilterFunction.class, "filter", Type.class, Block.class, MethodHandle.class);
+
+    private ArrayFilterFunction()
+    {
+        super(new Signature(
+                "filter",
+                FunctionKind.SCALAR,
+                ImmutableList.of(typeVariable("T")),
+                ImmutableList.of(),
+                parseTypeSignature("array(T)"),
+                ImmutableList.of(parseTypeSignature("array(T)"), parseTypeSignature("function(T,boolean)")),
+                false));
+    }
+
+    @Override
+    public boolean isHidden()
+    {
+        return true;
+    }
+
+    @Override
+    public boolean isDeterministic()
+    {
+        return false;
+    }
+
+    @Override
+    public String getDescription()
+    {
+        return "return array containing elements that match the given predicate";
+    }
+
+    @Override
+    public ScalarFunctionImplementation specialize(BoundVariables boundVariables, int arity, TypeManager typeManager, FunctionRegistry functionRegistry)
+    {
+        Type type = boundVariables.getTypeVariable("T");
+        return new ScalarFunctionImplementation(
+                false,
+                ImmutableList.of(false, false),
+                METHOD_HANDLE.bindTo(type),
+                isDeterministic());
+    }
+
+    public static Block filter(Type type, Block block, MethodHandle function)
+    {
+        int positionCount = block.getPositionCount();
+        BlockBuilder resultBuilder = type.createBlockBuilder(new BlockBuilderStatus(), positionCount);
+        for (int position = 0; position < positionCount; position++) {
+            Object input = readNativeValue(type, block, position);
+            Object keep;
+            try {
+                keep = (Object) function.invoke(input);
+            }
+            catch (Throwable throwable) {
+                throw Throwables.propagate(throwable);
+            }
+            if (TRUE.equals(keep)) {
```

Comment:
I'm not sure this has the right semantics. Since `NULL` means "unknown" if the result of the function is `NULL` shouldn't the output for that position be `NULL` instead of omitting it?